[distribtued.nn] Implement TorchScript-compatible RemoteModule API

xush6528 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#37139 [distribtued.nn] Implement TorchScript-compatible RemoteModule API**

See design doc in https://github.com/pytorch/pytorch/issues/37136

Differential Revision: [D20499658](https://our.internmc.facebook.com/intern/diff/D20499658/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20499658/)!

xush6528 committed:
[distribtued.nn] Implement TorchScript-compatible RemoteModule API

See design doc in https://github.com/pytorch/pytorch/issues/37136

Differential Revision: [D20499658](https://our.internmc.facebook.com/intern/diff/D20499658/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20499658/)!

[ghstack-poisoned]

xush6528 committed:
Update on "[WIP] [distribtued.nn] Implement TorchScript-compatible RemoteModule API"


See design doc in https://github.com/pytorch/pytorch/issues/37136

Differential Revision: [D20499658](https://our.internmc.facebook.com/intern/diff/D20499658/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20499658/)!

[ghstack-poisoned]

xush6528 committed:
Update on "[WIP] [distribtued.nn] Implement TorchScript-compatible RemoteModule API"


See design doc in https://github.com/pytorch/pytorch/issues/37136

Differential Revision: [D20499658](https://our.internmc.facebook.com/intern/diff/D20499658/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20499658/)!

[ghstack-poisoned]

xush6528 committed:
Update on "[WIP] [distribtued.nn] Implement TorchScript-compatible RemoteModule API"


See design doc in https://github.com/pytorch/pytorch/issues/37136

Differential Revision: [D20499658](https://our.internmc.facebook.com/intern/diff/D20499658/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20499658/)!

[ghstack-poisoned]

xush6528 committed:
Update on "[WIP] [distribtued.nn] Implement TorchScript-compatible RemoteModule API"


See design doc in https://github.com/pytorch/pytorch/issues/37136

Differential Revision: [D20499658](https://our.internmc.facebook.com/intern/diff/D20499658/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20499658/)!

[ghstack-poisoned]

mrshenli commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -0,0 +1,100 @@
+    args: Tuple = None,
+    kwargs: Dict[str, Any] = None,
+    global_unique_name: str = None,
+    module_interface_cls=None,
```
space around "="

xush6528 replied:
Let me annotate as Any type to align them.

mrshenli commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -0,0 +1,100 @@
+        of the module created by the module_creator.
+
+    Args:
+        to: str.
```
I assume this is user-facing? Shall we use the standard format: "arg_name (type): description"?

xush6528 replied:
Updating.

mrshenli commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -0,0 +1,100 @@
+    module_creator: Callable,
+    args: Tuple = None,
+    kwargs: Dict[str, Any] = None,
+    global_unique_name: str = None,
```
I assume this is used to pair the local and remote `module_cls`? Shall we add more comments or pointers about how the template work? And why we cannot let the local module_cls to hold an RRef to the remote module_cls to pair them?

mrshenli commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -0,0 +1,100 @@
+    kwargs = kwargs if kwargs is not None else {}
+    global_unique_name = (  # Assign a global name for the module to be created.
+        global_unique_name
+        if global_unique_name is not None
```
looks like we don't have a test to cover this case where `global_unique_name` is provided by the call site?

xush6528 replied:
Adding.

xush6528 replied:
Adding.

mrshenli commented on torch/testing/_internal/distributed/nn/api/remote_module_test.py:
```diff
@@ -0,0 +1,130 @@
+        if ModuleCreationMode.MODULE_CTOR_WITH_INTERFACE in modes:
+            remote_module = RemoteModule(
+                dst_worker_name,
+                MyModule,
+                args,
+                kwargs,
+                module_interface_cls=MyModuleInterface,
```
Why do we need a module creator and a module interface. What if these two do not match?

pritamdamania87 replied:
I feel we the user shouldn't have to pass in the ModuleInterface or a globally unique name.

xush6528 replied:
globally unique name is necessary for profiling purpose. We want to name all RemoteModules, so that we don't have to have a NamedEmbeddingBag wrapper and another NamedEmbedding wrapper. They can all use the RemoteModule, which takes a name.

xush6528 replied:
> Why do we need a module creator and a module interface. What if these two do not match?

Removing inferring module_interface_cls from module_creator to aovid confusion. Users always need to explicitly provide module_interface_cls if they want the RemoteModule to be torchscriptable.

mrshenli commented on aten/src/ATen/core/type.cpp:
```diff
@@ -659,8 +659,8 @@ bool ListType::isSubtypeOfExt(const TypePtr rhs_, std::ostream* why_not) const {
-   bool typesSame =
+bool TupleType::operator==(const Type& rhs) const {
+  bool typesSame =
        compare(rhs, [](const TypePtr a, const TypePtr b) { return *a == *b; });
```
also need to remove one space in this line?

xush6528 replied:
Reverting this change. It's unexpected formatting.

mrshenli commented on torch/distributed/nn/jit/instantiator.py:
```diff
@@ -0,0 +1,135 @@
+    out_path = os.path.join(
+        TEMPLATE_DIR_PATH, "instantiated", f"{generated_module_name}.py"
+    )
+    write(out_path, generated_code_text)
```
where and when do we delete the output file? or is it up to the application to delete the generated code file?

xush6528 replied:
See idea in https://github.com/pytorch/pytorch/pull/37139/files#r421161918.

mrshenli commented on torch/distributed/nn/jit/instantiator.py:
```diff
@@ -0,0 +1,135 @@
+    )
+    generated_code_text = remote_forward_template.substitute(env)
+    out_path = os.path.join(
+        TEMPLATE_DIR_PATH, "instantiated", f"{generated_module_name}.py"
```
Question, we are putting the generated file to the src folder? Which can potentially be in Python packages? I am not aware about the convention here, but does it make more sense to put this to /tmp folder? Otherwise, the number of different modules can keep growing in the package folder?

xush6528 replied:
> Question, we are putting the generated file to the src folder? Which can potentially be in Python packages? I am not aware about the convention here, but does it make more sense to put this to /tmp folder?

We can only put them into Python packages (folders with `__init__.py`), so that we can import them, right?
If we put them into `/tmp`, we can't import them.

> Otherwise, the number of different modules can keep growing in the package folder?
Since 
1) the template directory path is known to template instantiator 
and 2) RemoteModule only works on top of rpc, 
I am thinking about add a helper function that purge the template instances folder, and call this helper in `rpc.shutdown()`.

mrshenli commented on test/distributed/nn/jit/test_instantiator.py:
```diff
@@ -0,0 +1,55 @@
+
+    def test_instantiate_remote_module_template(self):
+        generated_module_name = f"_remote_module_{uuid.uuid4().hex}"
+        generated_module = instantiator.instantiate_remote_module_template(
```
If it is the application that should delete the generated code file, this test should delete that file here too?

xush6528 replied:
See the idea in https://github.com/pytorch/pytorch/pull/37139/files#r421161918, so that we can make the cleanup automatic and transparent.

mrshenli commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -0,0 +1,100 @@
+    Args:
+        to: str.
+        module_creator: Callable. The return type of the callable
+            must be a class decorated with @torch.jit.interface.
```
If this is user-facing doc, we will need examples.

xush6528 replied:
Adding an example.

pritamdamania87 commented on torch/distributed/nn/jit/instantiator.py:
```diff
@@ -0,0 +1,135 @@
+from torch.distributed.nn.jit.templates import dir_path as TEMPLATE_DIR_PATH
+
+
+def get_return_type_from_callable(callable_obj):
+    sig = inspect.signature(callable_obj)
+    return_annotation = sig.return_annotation
+    if return_annotation is inspect.Signature.empty:
+        return callable_obj
+    return return_annotation
```
I think we should have the following here:

1) Check if `callable_obj` is a class using `inspect.isclass()`. If this is true, we can just return callable_obj since this is the type of the Module that we're creating.
2) Check if callable_obj has return_annotation and if it does, return that as the type.
3) If 1) and 2) is not true, throw an exception mentioning that we couldn't determine the type of the Module and ask the user to add a return annotation.

xush6528 replied:
Updating.

pritamdamania87 commented on torch/testing/_internal/distributed/nn/api/remote_module_test.py:
```diff
@@ -0,0 +1,130 @@
+        pass
+
+
+class MyModule(nn.Module):
```
Can we verify something like `class MyScriptModule(torch.jit.ScriptModule):` works too?

xush6528 replied:
That's a deprecated API which is not used by us internally. I don't think we want to support that.

pritamdamania87 commented on torch/testing/_internal/distributed/nn/api/remote_module_test.py:
```diff
@@ -0,0 +1,130 @@
+            remote_module = RemoteModule(dst_worker_name, MyModule, args, kwargs)
+            yield remote_module
+
+    @dist_utils.dist_init
+    def test_create_remote_module(self):
+        if self.rank != 0:
+            return
+        dst_worker_name = dist_utils.worker_name((self.rank + 1) % self.world_size)
+        for remote_module in self._create_remote_module_iter(dst_worker_name):
+            pass
+
```
This test isn't necessary since we don't assert anything and other tests pretty much cover this codepath?

xush6528 replied:
Removing.

pritamdamania87 commented on torch/testing/_internal/distributed/nn/api/remote_module_test.py:
```diff
@@ -0,0 +1,130 @@
+import enum
+from typing import Tuple
+
+import torch
+import torch.testing._internal.dist_utils as dist_utils
```
Can we test that `forward` and `forward_async` can be called within a torchscript function?

xush6528 replied:
Adding.

pritamdamania87 commented on torch/testing/_internal/distributed/nn/api/remote_module_test.py:
```diff
@@ -0,0 +1,130 @@
+        return word, number, tensor
+
+
+def create_module(first_arg, first_kwarg=-1) -> MyModuleInterface:
+    return MyModule(first_arg, first_kwarg=first_kwarg)
```
Does this automatically create a ScriptModule due to the return annotation? Or do we need to do `torch.jit.script(MyModule(..))`.

If this already returns a ScriptModule, should we also have a test where returning `torch.jit.script(MyModule(...))` works as well?

xush6528 replied:
We still have to use `torch.jit.script(..)` to transform `MyModule `, which is a `torch.nn.Module`, into a ScriptModule, which is in Python type `torch.jit.RecursiveScriptModule`.

pritamdamania87 reviewed (changes requested):
Awesome work!

xush6528 commented on torch/distributed/nn/jit/instantiator.py:
```diff
@@ -0,0 +1,135 @@
+    out_path = os.path.join(
+        TEMPLATE_DIR_PATH, "instantiated", f"{generated_module_name}.py"
+    )
+    write(out_path, generated_code_text)
```
I tried `exec`. It turned out that JIT frontend opens a file to parse the code text. So I changed the PR to generate a file instead.

xush6528 commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -0,0 +1,100 @@
+    # Default arguments preperation.
+    args = args if args is not None else ()
+    kwargs = kwargs if kwargs is not None else {}
+    global_unique_name = (  # Assign a global name for the module to be created.
+        global_unique_name
+        if global_unique_name is not None
+        else _gen_global_unique_name()
+    )
+
+    # Infer module_interface type.
+    module_interface_cls = (
+        module_interface_cls
+        if module_interface_cls is not None
+        else instantiator.get_return_type_from_callable(module_creator)
+    )
```
Discussed with @wanchaol when doing the same thing for RRef constructor. Due to multi-inheritance and per-instance concrete type, there is no easy way to infer the interface type of a script module type. The most straightforward way is to ask user to provide the interface which includes the methods they will use.

xush6528 commented on test/distributed/nn/jit/test_instantiator.py:
```diff
@@ -0,0 +1,55 @@
+#!/usr/bin/env python3
+import unittest
+import uuid
```
This is what formatter gave me.

xush6528 commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -0,0 +1,100 @@
+
+def _instantiate_template(global_unique_name, module_interface_cls, is_scriptable):
+    # Generate the template instance name.
+    generated_module_name = f"_remote_module_{global_unique_name}"
```
`__name__` will have naming collision, let me use qualified name with "." replace by "_".

xush6528 replied:
Multiple processes writing to the same file corrupts the file.
If we do want to do that, I need to introduce file lock.
https://docs.python.org/3/library/fcntl.html

xush6528 commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -0,0 +1,100 @@
+    return rpc.RRef(script_module, module_interface_cls)
+
+
+def RemoteModule(
```
This won't work. We can't dynamically annotate the signature of this Remote class.

mrshenli replied:
Is this a temporary limitation of JIT, or some fundamental constraint?

mrshenli replied:
If this has to be a function, shall we call this `create_remote_model` instead? Or is it possible to have a `class RemoteModel` with a separate `_create_remote_model` function. Then in `RemoteModel.__new__`, we use `_create_remote_model` to create a remote module instance?

xush6528 replied:
This is how JIT works, it must know the concrete code with type annotation.

I can try this idea.

xush6528 committed:
Update on "[WIP] [distribtued.nn] Implement TorchScript-compatible RemoteModule API"


See design doc in https://github.com/pytorch/pytorch/issues/37136

Differential Revision: [D20499658](https://our.internmc.facebook.com/intern/diff/D20499658/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20499658/)!

[ghstack-poisoned]

xush6528 commented:
Addressed some comments.

TODOs,
- Debug OSS importing ModuleNotFound error.
- Discuss about generated file output path.
- Write up docstrings.

xush6528 commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -0,0 +1,100 @@
+
+    Args:
+        to: str.
+        module_creator: Callable. The return type of the callable
+            must be a class decorated with @torch.jit.interface.
```
I'm removing this weird requirement to avoid confusion to users.

xush6528 committed:
Update on "[WIP] [distribtued.nn] Implement TorchScript-compatible RemoteModule API"


See design doc in https://github.com/pytorch/pytorch/issues/37136

Differential Revision: [D20499658](https://our.internmc.facebook.com/intern/diff/D20499658/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20499658/)!

[ghstack-poisoned]

xush6528 commented:
For the exception
```bash
May 16 06:24:14 ======================================================================
May 16 06:24:14 ERROR [0.004s]: test_instantiate_remote_module_template (__main__.TestInstantiator)
May 16 06:24:14 ----------------------------------------------------------------------
May 16 06:24:14 Traceback (most recent call last):
May 16 06:24:14   File "distributed/nn/jit/test_instantiator.py", line 56, in test_instantiate_remote_module_template
May 16 06:24:14     MyModuleInterface, True  # is_scriptable
May 16 06:24:14   File "/opt/conda/lib/python3.6/site-packages/torch/distributed/nn/jit/instantiator.py", line 113, in instantiate_remote_module_template
May 16 06:24:14     os.path.join(TEMPLATE_DIR_PATH, "remote_module.py.template")
May 16 06:24:14   File "/opt/conda/lib/python3.6/site-packages/torch/distributed/nn/jit/code_template.py", line 31, in from_file
May 16 06:24:14     with open(filename, 'r') as f:
May 16 06:24:14 FileNotFoundError: [Errno 2] No such file or directory: '/opt/conda/lib/python3.6/site-packages/torch/distributed/nn/jit/templates/remote_module.py.template'
```

Only the command `python setup.py install` can reproduce this error, because for Python packaging tool, "setuptool",
1) It copies files into the path, "/opt/conda/lib/python3.6/site-packages/torch".
2) It only copies files end with ".py".

The command `python setup.py develop` does not reproduce this error, because
it create a symbolic link at "/opt/conda/lib/python3.6/site-packages/torch", pointing to torch src code "torch" sub-folder path.
So it doesn't matter to all non-".py" files.

The solution is to add the needed template file to the "package_data" list. "package_data" is an argument for the entry API, `setup(...)` function, of setuptool.

xush6528 committed:
Update on "[WIP] [distribtued.nn] Implement TorchScript-compatible RemoteModule API"


See design doc in https://github.com/pytorch/pytorch/issues/37136

Differential Revision: [D20499658](https://our.internmc.facebook.com/intern/diff/D20499658/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20499658/)!

[ghstack-poisoned]

xush6528 commented:
Updates
- Resolve the ModuleNotFound.
- Resolved the FileNotFoundError.
- Add cleanup generated module files.

xush6528 commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -0,0 +1,100 @@
+    module_interface_cls=None,
+):
+    """
+        The returned RemoteModule instance will have auto-generated
+        forward_async / forward methods, check the implementation in
+        remote_module.py.template,
+        the arguments they take are the same as the forward method
+        of the module created by the module_creator.
```
Thanks! Adding this.

xush6528 commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -0,0 +1,100 @@
+        of the module created by the module_creator.
+
+    Args:
+        to: str.
```
Doing it.

xush6528 committed:
Update on "[distribtued.nn] Implement TorchScript-compatible RemoteModule API"


See design doc in https://github.com/pytorch/pytorch/issues/37136

Differential Revision: [D20499658](https://our.internmc.facebook.com/intern/diff/D20499658/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20499658/)!

[ghstack-poisoned]

pritamdamania87 commented on torch/distributed/nn/jit/templates/remote_module.py.template:
```diff
@@ -0,0 +1,58 @@
+        super().__init__()
+        self.module_rref = module_rref
+        self.is_scriptable = is_scriptable
+        self.global_unique_name = global_unique_name
```
We can get rid of global_unique_name since it is no longer used?

xush6528 replied:
I want to make RemoteModule a convenient wrapper.

This field will be used by post processing. For example, checkpoint and re-sharding will need to locate the shard wrapped by RemoteModule.

pritamdamania87 commented on torch/distributed/nn/jit/instantiator.py:
```diff
@@ -0,0 +1,151 @@
+        **jit_decorator_str_map,
+    )
+    generated_code_text = remote_forward_template.substitute(env)
+    out_path = os.path.join(
+        INSTANTIATED_TEMPLATE_DIR_PATH, f"{generated_module_name}.py"
+    )
```
Can we just use python tempfile (https://docs.python.org/3/library/tempfile.html) for this and rely on python to delete these files?

xush6528 replied:
Then, we'll need to create a `__init__.py` there and add it to `sys.path`, does that sound good to you?

xush6528 replied:
This won't work. 

Since the TorchScript function `_remote_forward` is referenced by qualifed_name, they must be put at the same python module path on both client and server sides.
However, `tempfile` always generates a random file name.

pritamdamania87 commented on torch/distributed/nn/jit/code_template.py:
```diff
@@ -0,0 +1,73 @@
+
+import re
+
+# match $identifier or ${identifier} and replace with value in env
+# If this identifier is at the beginning of whitespace on a line
+# and its value is a list then it is treated as
+# block subsitution by indenting to that depth and putting each element
+# of the list on its own line
+# if the identifier is on a line starting with non-whitespace and a list
+# then it is comma separated ${,foo} will insert a comma before the list
+# if this list is not empty and ${foo,} will insert one after.
```
Why do we need to build this ourselves, can't we just use python's format(https://docs.python.org/3.4/library/functions.html#format) to appropriately replace the correct env variables?

xush6528 replied:
Replace if with python `str.format(..)`.

pritamdamania87 commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -0,0 +1,129 @@
+    )
+
+    # Instantiate template on local side.
+    remote_module_cls, _remote_module_interface_cls = _instantiate_template(module_interface_cls, is_scriptable)
```
`_remote_module_interface_cls` is never used?

xush6528 replied:
Removing.

pritamdamania87 commented on torch/distributed/nn/jit/templates/remote_module.py.template:
```diff
@@ -0,0 +1,58 @@
+    def __init__(self, module_rref, is_scriptable, global_unique_name):
+        super().__init__()
+        self.module_rref = module_rref
+        self.is_scriptable = is_scriptable
```
self.is_scriptable is not used either, do we need this?

xush6528 replied:
It's useful for users. As I mentioned, we can't script RemoteModule for users, because scripted modules can't be send over wire. Users will rely on this flag to decide whether to script the remote module returned by this API.

pritamdamania87 commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -0,0 +1,129 @@
+    args: Tuple = None,
+    kwargs: Dict[str, Any] = None,
+    global_unique_name: str = None,
+    module_interface_cls: Any = None,
```
I was wondering if we can do the following to avoid using `module_interface_cls`. We support only the following cases:

```
// 1. Use nn.Module directly.
RemoteModule(module_creator=nn.Linear())
NOTE: I think we can also easily support the case where module_creator itself is a type of ScriptModule.

// 2. Use nn.Module as part of function.
def my_module_creator(): -> nn.Module
  return MyModule()

RemoteModule(module_creator=my_module_creator())

// 3. Use nn.Module as ScriptModule.
def my_script_module_creator(): -> ScriptModule
  return torch.jit.script(MyModule())

RemoteModule(module_creator=my_script_module_creator())
```

In each of these cases, we can inspect either the class or the return type of the creator to determine what `module_interface_cls` should be. If the user doesn't specify a return type annotation, we just raise an error and ask them to provide one?

Also, I was chatting with @wanchaol offline to see if we can come up with options to make this cleaner. I think it might make sense to have a 30min meeting next week to discuss a few ideas around this.

xush6528 replied:
RemoteModule is relying on local RRef constructor API. 

A separate PR later should clean up local RRef constructor API and wrap type info when passing and pickling RRefs.

This will need careful design.

pritamdamania87 commented on torch/distributed/nn/jit/instantiator.py:
```diff
@@ -0,0 +1,151 @@
+        return module_interface_cls
+
+    # Otherwise, if module_creator is a class ctor, use it.
+    if inspect.isclass(module_creator):
+        return module_creator
```
Shouldn't we also check that the class is also either of type nn.Module or ScriptModule?

xush6528 replied:
Good point! 
Should be nn.Module or RecursiveScriptModule.

xush6528 commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -0,0 +1,129 @@
+        remote module.
+
+    Example::
+        Then run the following code in two different processes:
```
Done.

xush6528 commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -0,0 +1,129 @@
+
+    Returns:
+        A user :class:`~nn.Module` delegate instance of the remote module created by
+        module_creator, it has a blocking forward method and an asyncrounous
```
Done

xush6528 commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -0,0 +1,129 @@
+            must be a class decorated with @torch.jit.interface.
+        args (Sequence): args to be passed to module_creator.
+        kwargs (Dict): kwargs to be passed to module_creator.
+        global_unique_name (str): The unique name of the created RemoteModule,
+            useful for profiling purpose. If not provided, a UUID4 will
+            be generated as its name.
```
Users will need to specify. For checkpointing / transfer learning / re-sharding purpose, users will have a systematic way of generating Module name. Similar to caffe2 blob name generating.

pritamdamania87 replied:
Could you elaborate a bit here in terms of how this globally unique name would be used by users? Current nn.modules don't have a name associated with them. What sort of APIs do you have in mind which could benefit from global_unique_name? My feeling is that we can always add this later if needed instead (since it is optional anyways) of adding it from the get go to the API.

xush6528 replied:
Internally we added this field for profiling purpose.
https://fburl.com/diffusion/8vnlbm5s

We name the remote modules when we materialize them. 
https://fburl.com/diffusion/mdqhxu4i

Now, since there is no torchscript compatible `RemoteModule` wrapper API. I have to add it to the concrete module classes. For example, D20166971.

xush6528 commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -0,0 +1,129 @@
+        executed on the remote node.
+        It takes care of autograd recording to ensure the backward pass propogates
+        gradients back to the corresponding remote module.
+        The arguments of foward_async and foward they take are the same
+        as the forward method of the module returned by the module_creator.
```
Done

xush6528 commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -0,0 +1,129 @@
+        global_unique_name (str): The unique name of the created RemoteModule,
+            useful for profiling purpose. If not provided, a UUID4 will
+            be generated as its name.
+        module_interface_cls (type): The TorchScript interface type for the module
+            to be created. If not provided, the module is not torchscript-able.
```
Done for all optional args.

xush6528 commented on torch/distributed/nn/jit/instantiator.py:
```diff
@@ -0,0 +1,151 @@
+    if inspect.isclass(module_creator):
+        return module_creator
+
+    raise ValueError(
+        "Can't infer module_interface_cls, "
+        "because neither module_interface_cls nor module_creator return type is provided."
+    )
```
This behavior is not a common practice for Python code. Removed this behavior.

mrshenli commented on torch/csrc/jit/python/python_ir.cpp:
```diff
@@ -780,6 +780,12 @@ void initPythonIRBindings(PyObject* module_) {
+          [](InterfaceType& self, const std::string& name) {
+            return self.getMethod(name);
+          },
+          py::return_value_policy::reference)
```
This means the lifetime of the returned `method` object is managed by C++? And it will be alive until the CU is gone?

xush6528 replied:
Correct. 
Notice `self.getMethod(name)` returns a raw pointer. 

If not passed by reference, pybind11 creates a unique_ptr wrapping the raw pointer.

So Python GC will `delete` the raw pointer when destructing the unique_ptr, causing later accesses from C++ world using a dangling pointer.

mrshenli commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -0,0 +1,129 @@
+        executed on the remote node.
+        It takes care of autograd recording to ensure the backward pass propogates
+        gradients back to the corresponding remote module.
+        The arguments of foward_async and foward they take are the same
```
typo:

foward_async -> forward_async
foward -> forward

remove "they take"?

xush6528 replied:
Updating.

mrshenli commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -0,0 +1,129 @@
+        It takes care of autograd recording to ensure the backward pass propogates
+        gradients back to the corresponding remote module.
+        The arguments of foward_async and foward they take are the same
+        as the forward method of the module returned by the module_creator.
```
double quote for `module_creator`, `foward_async` and `forward`

xush6528 replied:
Adding.

mrshenli commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -0,0 +1,129 @@
+    Returns:
+        A user :class:`~nn.Module` delegate instance of the remote module created by
+        module_creator, it has a blocking forward method and an asyncrounous
+        foward_aynsc that returns a future of the foward call on the user created
```
s/foward/forward/g 

let's do this for the entire file

xush6528 replied:
Doing it.

mrshenli commented on torch/distributed/nn/jit/instantiator.py:
```diff
@@ -0,0 +1,151 @@
+        arrow_and_return_type_str = ""
+        arrow_and_future_return_type_str = ""
+
+    jit_decorator_str_map = dict(
+        jit_script_decorator="@torch.jit.script",
+        jit_export_decorator="@torch.jit.export",
+        jit_interface_decorator="@torch.jit.interface",
+    )
+    if is_scriptable is False:
+        jit_decorator_str_map = {
+            key: "" for key, value in jit_decorator_str_map.items()
+        }
```
move `jit_decorator_str_map` before the above if-else, and merge this if into the `else` clause.

xush6528 replied:
I have split the scripted and non scripted code paths in latest version.

mrshenli commented on test/run_test.py:
```diff
@@ -70,6 +70,8 @@
     'test_jit_fuser_te',
     'test_tensorexpr',
+    'distributed/nn/jit/test_instantiator',
+    'distributed/nn/api/test_remote_module_spawn',
```
should this also be added to `SLOW_TESTS`?

xush6528 replied:
Good point! Adding.

xush6528 commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -0,0 +1,129 @@
+            to be created. If not provided, the module is not torchscript-able.
+
+    Returns:
+        A user :class:`~nn.Module` delegate instance of the remote module created by
+        module_creator, it has a blocking forward method and an asyncrounous
```
Updating.

xush6528 commented on torch/distributed/nn/jit/instantiator.py:
```diff
@@ -0,0 +1,151 @@
+from torch.distributed.nn.jit.templates.instantiated import INSTANTIATED_TEMPLATE_DIR_PATH
+
+
+def infer_module_interface_cls(module_creator, module_interface_cls):
+    # If module_interface_cls is provided, use it.
+    if module_interface_cls is not None:
+        return module_interface_cls
+
+    # Otherwise, if module_creator is a class ctor, use it.
+    if inspect.isclass(module_creator):
+        return module_creator
+
+    raise ValueError(
+        "Can't infer module_interface_cls, "
+        "because neither module_interface_cls nor module_creator return type is provided."
+    )
```
Adding.

xush6528 commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -0,0 +1,129 @@
+
+    Arguments:
+        to (str or WorkerInfo): id or name of the destination worker.
+        module_creator (Callable): The return type of the callable
+            must be a class decorated with @torch.jit.interface.
```
Updating.

xush6528 commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -0,0 +1,129 @@
+        executed on the remote node.
+        It takes care of autograd recording to ensure the backward pass propogates
+        gradients back to the corresponding remote module.
+        The arguments of foward_async and foward they take are the same
+        as the forward method of the module returned by the module_creator.
```
Adding.

xush6528 committed:
Update on "[distribtued.nn] Implement TorchScript-compatible RemoteModule API"


See design doc in https://github.com/pytorch/pytorch/issues/37136

Differential Revision: [D20499658](https://our.internmc.facebook.com/intern/diff/D20499658/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20499658/)!

[ghstack-poisoned]

xush6528 committed:
Update on "[distribtued.nn] Implement TorchScript-compatible RemoteModule API"


See design doc in https://github.com/pytorch/pytorch/issues/37136

Differential Revision: [D20499658](https://our.internmc.facebook.com/intern/diff/D20499658/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20499658/)!

[ghstack-poisoned]

xush6528 committed:
Update on "[distribtued.nn] Implement TorchScript-compatible RemoteModule API"


See design doc in https://github.com/pytorch/pytorch/issues/37136

Differential Revision: [D20499658](https://our.internmc.facebook.com/intern/diff/D20499658/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20499658/)!

[ghstack-poisoned]

xush6528 committed:
Update on "[distribtued.nn] Implement TorchScript-compatible RemoteModule API"


See design doc in https://github.com/pytorch/pytorch/issues/37136

Differential Revision: [D20499658](https://our.internmc.facebook.com/intern/diff/D20499658/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20499658/)!

[ghstack-poisoned]

xush6528 committed:
Update on "[distribtued.nn] Implement TorchScript-compatible RemoteModule API"


See design doc in https://github.com/pytorch/pytorch/issues/37136

Differential Revision: [D20499658](https://our.internmc.facebook.com/intern/diff/D20499658/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20499658/)!

[ghstack-poisoned]

mrshenli commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -0,0 +1,245 @@
+    """
+        A RemoteModule instance can only be created after RPC initialization.
+        It creates a user-specified module on a specified remote node.
+        It behaves like a regular nn.Module except that the ``forward`` method is
```
double quote `nn.Module`

xush6528 replied:
Updated.

mrshenli commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -0,0 +1,245 @@
+    return remote_module
+
+
+def RemoteModule(
```
I recall there was a discussion regarding whether this should be a `class` and annotate it's `__new__` method. Is it because that didn't work, so that we still make this a function?

xush6528 replied:
Going with this idea instead. https://github.com/pytorch/pytorch/pull/37139#discussion_r434201926

mrshenli commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -0,0 +1,245 @@
+    """
+        A RemoteModule instance can only be created after RPC initialization.
+        It creates a user-specified module on a specified remote node.
+        It behaves like a regular nn.Module except that the ``forward`` method is
```
double quote `nn.Module`

xush6528 replied:
Updated.

mrshenli commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -0,0 +1,245 @@
+
+        For example, if ``module_creator`` returns an instace of ``nn.Linear``,
+        that has ``forward`` method signature, ``def forward(input: Tensor) -> Tensor:``,
+        the generated `RemoteModule`` will have 2 methods in signature of
```
missing one quote before `RemoteModule`.

xush6528 replied:
Fixed.

mrshenli commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -0,0 +1,245 @@
+    Arguments:
+        to (str or WorkerInfo): id or name of the destination worker.
+        module_creator (Callable): A ``module_creator`` could be
+                1. A type object that is subclass of ``nn.Module``.
```
Is it intentional to have 8-space indent here?

xush6528 replied:
Fixing it.

mrshenli commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -0,0 +1,245 @@
+    kwargs: Dict[str, Any] = None,
+    global_unique_name: str = None,
+):
+    """
```
If we aim to expose this to public API docs in v1.6, let's generate doc htmls and check if the format is as expected.

mrshenli commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -0,0 +1,245 @@
+    Returns:
+        A remote module instance which wraps the :class:`~nn.Module` created by the
+        user-provided ``module_creator``, it has a blocking ``forward`` method and an asynchronous
+        ``forward`` that returns a future of the ``forward`` call on the user created
```
the first one should be ``forward_async``?

xush6528 replied:
Updated.

xush6528 committed:
Update on "[distribtued.nn] Implement TorchScript-compatible RemoteModule API"


See design doc in https://github.com/pytorch/pytorch/issues/37136

Differential Revision: [D20499658](https://our.internmc.facebook.com/intern/diff/D20499658/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20499658/)!

[ghstack-poisoned]

wanchaol commented on torch/distributed/nn/jit/instantiator.py:
```diff
@@ -0,0 +1,162 @@
+    # to be noticed by the import system.
+    importlib.invalidate_caches()
+    generated_module = importlib.import_module(
+        f"torch.distributed.nn.jit.templates.instantiated.{generated_module_name}"
```
this template seems not only getting used by jit, also getting used by nn, so maybe don't put it in the jit foler?

xush6528 replied:
I put it separate because this TorchScript support is an extension to `nn.RemoteModule`.
If the created RemoteModule does not to be scripted, code in this file is not called.

Opening a separate folder because the template initiation tend to become more complex over time, if we want to support generating methods beyond "forward".

wanchaol commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -0,0 +1,245 @@
+        ``def forward(input: Tensor) -> Tensor:`` and
+        ``def forward_async(input: Tensor) -> Future[Tensor]:``.
+
+    Arguments:
```
I feel like a nicer API for RemoteModule could be

```
class RemoteModule(nn.Module):
	def __init__(to: str, local_module_cls, args, kwargs):
        self.to = to
        self.mod_rref = rpc.remote(to, local_module_cls, args, kwargs)

	def forward():
		# code for forward

	def forward_async():
		# code for async forward
```

For TorchScript support, we temporarily do

```
class _RemoteModule(nn.Module):
	def __init__(to: str, local_module_cls, args, kwargs, _module_interface_cls):
        self.to = to
        # call instantiator to instantiate the forward/forward_async methods from interface
		instantiated_forward_method = instantiator.instantiate_forward(interface_cls)
		instantiated_forward_async_method = instantiator.instatiate_forward_async(interface_cls)
		setattr(self, instantiated_forward_method)
		setattr(self, instantiated_forward_async_method)

xush6528 replied:
Not planning to do it. This way we will maintain two duplicate code. Instead, I only want to keep one ground truth in the template that can generate for both Python (without method signatures) and TorchScript(with method signatures).

xush6528 replied:
Updating
- Make both Python API and TorchScript API a class inherit from `nn.Module`.
- Simplify template to generate methods only, instead of generatin a whole class.
- Dynamically attach generated methods to the instance in `__init__`.

wanchaol reviewed (commented):
Great progress! I have some comments on the API that we provide to the user, and some suggestions on how do we do the codegen according to module_interface_cls. Please see inline comments.

pritamdamania87 commented on torch/distributed/rpc/api.py:
```diff
@@ -260,6 +261,14 @@ def shutdown(graceful=True):
         # resolved.
         _cleanup_python_rpc_handler()
         _reset_current_rpc_agent()
+    try:
+        # This API removes files. There is chance of failure doing IO.
+        instantiator.cleanup_generated_modules()
+    except Exception:
+        import traceback
+
+        print("Encountered exception in instantiator.cleanup_generated_modules()")
+        traceback.print_exc()
```
Can we do this using `atexit` instead: https://docs.python.org/3.6/library/atexit.html? This is more reliable since the user might forget to call shutdown() or shutdown() might not be called due to a bug in the user's code.

xush6528 replied:
Good idea! Didn't know this mechanism before. Updating.

pritamdamania87 commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -0,0 +1,245 @@
+        ``def forward_async(input: Tensor) -> Future[Tensor]:``.
+
+    Arguments:
+        to (str or WorkerInfo): id or name of the destination worker.
```
Nit: just noticed this, maybe we should call this `on` instead and mention in the docs `id or name of the worker on which this module should be created`. `on` is probably more accurate here since we're creating something **on** a worker instead of sending something **to** a worker when using rpc_async etc.

xush6528 replied:
Sounds better. Updating.

pritamdamania87 commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -0,0 +1,245 @@
+        remote_module_cls = _NON_SCRIPTABLE_REMOTE_MODULE_CLS
+
+        # Create the module on the remote side.
+        module_rref = rpc.remote(to, module_creator, args, kwargs)
```
Shouldn't we validate that `module_creator` actually returns an nn.Module type? Wouldn't this just work if module_creator doesn't return nn.Module and then when the user calls `forward` it would fail with a cryptic error message then?

I feel we should validate at initialization type that we did in fact create a remote module on the destination.

xush6528 replied:
Good point! Adding the check.

pritamdamania87 commented on torch/testing/_internal/distributed/nn/api/remote_module_test.py:
```diff
@@ -0,0 +1,205 @@
+    _gen_global_unique_name,
+    _RemoteModule,
+)
+from torch.testing._internal.distributed.rpc.rpc_agent_test_fixture import (
```
Can we test using invalid module_creators throw an appropriate exception? Examples:

1. module_creator is a class that is not type of nn.Module.
2. module_creator returns an object that is not type of nn.Module.

xush6528 replied:
Adding corresponding tests.

xush6528 commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -0,0 +1,129 @@
+    return rpc.RRef(script_module, module_interface_cls)
+
+
+def RemoteModule(
```
Yes, it is. It's an object of class in "torch/distributed/nn/jit/templates/remote_module_template.py". line 20.

xush6528 commented on torch/distributed/nn/jit/instantiator.py:
```diff
@@ -0,0 +1,154 @@
+def get_arg_return_types_from_interface(module_interface):
+    assert getattr(
+        module_interface, "__torch_script_interface__", False
+    ), "Expect a class decorated by @torch.jit.interface."
```
Updating

xush6528 commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -0,0 +1,245 @@
+    args: Tuple = None,
+    kwargs: Dict[str, Any] = None,
+    global_unique_name: str = None,
+    module_interface_cls: Any = None,
```
Updating

xush6528 commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -0,0 +1,245 @@
+        >>> from torch import nn, Tensor
+        >>> from torch.distributed.nn.api.remote_module import RemoteModule
+        >>>
+        >>> @torch.jit.interface
```
Removing

xush6528 commented on torch/distributed/nn/jit/instantiator.py:
```diff
@@ -0,0 +1,162 @@
+    cu = torch.jit._python_cu
+    module_interface_c = cu.get_interface(qualified_name)
+    assert (
+        "forward" in module_interface_c.getMethodNames()
```
Yes, for this version of RemoteModule. I only support code generation for "forward". Will consider rest methods when the RRef with type info is there.

xush6528 commented on torch/distributed/nn/jit/templates/remote_module_template.py:
```diff
@@ -0,0 +1,67 @@
+
+
+{jit_interface_decorator}
+class _RemoteModuleInterface:
```
It's not used in this version. I noticed we can't import interface class from another module. JIT does not recognize it. We can only define interface in the same module where it is used.

xush6528 commented on torch/distributed/nn/jit/instantiator.py:
```diff
@@ -0,0 +1,162 @@
+        print("Skipped writing {}".format(out_path))
+
+
+def _do_instantiate_remote_module_template(generated_module_name, str_dict):
```
Answered in https://github.com/pytorch/pytorch/pull/37139#discussion_r418745513. It's a JIT limitation.

xush6528 committed:
Update on "[distribtued.nn] Implement TorchScript-compatible RemoteModule API"


See design doc in https://github.com/pytorch/pytorch/issues/37136

Differential Revision: [D20499658](https://our.internmc.facebook.com/intern/diff/D20499658/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20499658/)!

[ghstack-poisoned]

xush6528 committed:
Update on "[distribtued.nn] Implement TorchScript-compatible RemoteModule API"


See design doc in https://github.com/pytorch/pytorch/issues/37136

Differential Revision: [D20499658](https://our.internmc.facebook.com/intern/diff/D20499658/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20499658/)!

[ghstack-poisoned]

xush6528 committed:
Update on "[distribtued.nn] Implement TorchScript-compatible RemoteModule API"


See design doc in https://github.com/pytorch/pytorch/issues/37136

Differential Revision: [D20499658](https://our.internmc.facebook.com/intern/diff/D20499658/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20499658/)!

[ghstack-poisoned]

wanchaol reviewed (approved):
looks good to me, thanks for making the API better! left some minor comments and questions.

xush6528 commented on torch/distributed/nn/jit/templates/remote_module_template.py:
```diff
@@ -0,0 +1,47 @@
+
+REMOTE_MODULE_TEMPLATE = """from typing import Tuple, Any
+
+import torch
```
Only `nn` is not used. 

But for other types, the ones in `typing` module, and `torch.Tensor` type, we need them. When we generate user-provided `forward` method, these types might be referenced.

xush6528 commented on torch/testing/_internal/distributed/nn/api/remote_module_test.py:
```diff
@@ -0,0 +1,244 @@
+import torch.testing._internal.dist_utils as dist_utils
+from torch import Tensor, nn
+from torch._jit_internal import Future
+from torch.distributed.nn.api.remote_module import (
```
Updating.

xush6528 commented on torch/distributed/nn/api/remote_module.py:
```diff
@@ -0,0 +1,262 @@
+            f"but it returns an instance of {type(module)}."
+        )
+    if module_interface_cls is not None:
+        module = torch.jit.script(module)
```
I am choosing to compile it for users, thus the cognitive cost of this API could be monimal.
Updating the docs and arg name to this idea.

xush6528 committed:
Update on "[distribtued.nn] Implement TorchScript-compatible RemoteModule API"


See design doc in https://github.com/pytorch/pytorch/issues/37136

Differential Revision: [D20499658](https://our.internmc.facebook.com/intern/diff/D20499658/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20499658/)!

[ghstack-poisoned]

xush6528 committed:
Update on "[distribtued.nn] Implement TorchScript-compatible RemoteModule API"


See design doc in https://github.com/pytorch/pytorch/issues/37136

Differential Revision: [D20499658](https://our.internmc.facebook.com/intern/diff/D20499658/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20499658/)!

[ghstack-poisoned]

xush6528 committed:
Update on "[distribtued.nn] Implement TorchScript-compatible RemoteModule API"


See design doc in https://github.com/pytorch/pytorch/issues/37136

Differential Revision: [D20499658](https://our.internmc.facebook.com/intern/diff/D20499658/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20499658/)!

[ghstack-poisoned]
